cmd/zoekt-sourcegraph-indexserver/backoff_test.go

```diff
@@ -0,0 +1,249 @@
+package main
+
+import (
+	"github.com/sourcegraph/log/logtest"
+	"testing"
+	"time"
+)
+
+func TestQueue_BackoffOnFail(t *testing.T) {
+	backoffDuration := 1 * time.Millisecond
+	maxBackoffDuration := backoffDuration * 2
+
+	queue := NewQueue(backoffDuration, maxBackoffDuration, logtest.Scoped(t))
+	opts := IndexOptions{RepoID: 1, Name: "foo"}
+
+	queue.AddOrUpdate(opts)
+
+	// Empty queue
+	for ok := true; ok; _, ok = queue.Pop() {
+	}
+
+	queue.SetIndexed(opts, indexStateFail)
+
+	bumpTime := time.Now()
+	queue.Bump([]uint32{opts.RepoID})
+
+	// item is disallowed from being pushed to heap during backoff period
+	if item, ok := queue.Pop(); ok {
+		qi := queue.items[item.RepoID]
+		if qi.backoff.backoffUntil.Before(bumpTime) {
+			t.Errorf("backoffDuration already passed before first attempt to push item to heap in Bump(). Increase backoffDuration for the Queue. backoffDuration: %s. maxBackoffDuration: %s.",
+				backoffDuration, maxBackoffDuration)
+		} else {
+			t.Fatal("queue should be empty")
+		}
+	}
+}
+
+func TestQueue_BackoffAllowAfterDuration(t *testing.T) {
+	backoffDuration := 1 * time.Millisecond
+	maxBackoffDuration := backoffDuration * 2
+
+	queue := NewQueue(backoffDuration, maxBackoffDuration, logtest.Scoped(t))
+	opts := IndexOptions{RepoID: 1, Name: "foo"}
+
+	queue.AddOrUpdate(opts)
+
+	// Empty queue
+	for ok := true; ok; _, ok = queue.Pop() {
+	}
+
+	queue.SetIndexed(opts, indexStateFail)
+
+	if _, ok := queue.Pop(); ok {
+		t.Fatal("queue should be empty after SetIndexed")
+	}
+
+	time.Sleep(backoffDuration * 2)
```

Comment:
this should be fine, but just to be safe I'd sleep for a little bit longer.

Reply:
sleep increased